moneriote.py

```diff
@@ -1,303 +1,557 @@
+            obj.valid = True
+        return obj
+
+
+class Moneriote:
+    def __init__(self,
+                 cf_domain_name: str,
+                 cf_subdomain_name: str,
+                 cf_dns_api_key: str, cf_dns_api_email: str,
+                 cf_max_records: int = 5,
+                 md_address: str = '127.0.0.1',
+                 md_port: int = 18082,
+                 md_auth: str = 'not:used',
+                 md_path: str = 'monerod.exe',
+                 md_height_discovery_method: str = 'xmrchain'):
+        """See config.ini for documentation on the params."""
+        self.md_path = md_path
+        self.md_daemon_addr = md_address
+        self.md_daemon_port = md_port
+        self.md_daemon_auth = md_auth
+        if md_height_discovery_method not in ['xmrchain', 'monerod', 'both']:
+            log_err('bad height_discovery_method option')
+            sys.exit()
+        self.md_height_discovery_method = md_height_discovery_method
+
+        self.cf_api_base = 'https://api.cloudflare.com/client/v4/zones'
+
+        # zone_id will be detected via Cloudflare API
+        self.cf_zone_id = None
+        self.cf_dns_api_key = cf_dns_api_key
+        self.cf_dns_api_email = cf_dns_api_email
+
+        # node
+        self.cf_subdomain_name = cf_subdomain_name
+
+        # example.com
+        self.cf_domain_name = cf_domain_name
+
+        # node.example.com
+        self.cf_fulldomain_name = '%s.%s' % (cf_subdomain_name, cf_domain_name)
+
+        # the maximum amount of A records to add
+        self.cf_max_records = cf_max_records
+        self._cf_headers = {
+            'Content-Type': 'application/json',
+            'X-Auth-Email': self.cf_dns_api_email,
+            'X-Auth-Key': self.cf_dns_api_key,
+            'User-Agent': random_user_agent()
+        }
+
+        # how many servers we should scan at once
+        self._concurrent_scans = 20
+        # 10 minutes
+        self._loop_interval = 60 * 10
+        # default Monero RPC port
+        self._m_rpc_port = 18089
+        self._blockchain_height = None
+
+        if not os.path.isfile(self._path_nodes_cache):
+            log_msg("Auto creating \'%s\'" % self._path_nodes_cache)
+            f = open(self._path_nodes_cache, 'a')
+            f.write('[]')
+            f.close()
+
+    @classmethod
+    def from_config(cls):
+        if not os.path.isfile('config.ini'):
+            log_err("config.ini missing")
+            sys.exit()
+
+        config = configparser.ConfigParser()
+        config.read('config.ini')
+
+        def try_cast(val):
+            if val.isdigit():
+                return int(val)
+            if val.lower() in ['true', 'false']:
+                return bool(val)
+            return val
+
+        cfg_monero = {'md_%s' % k: try_cast(v) for k, v in config._sections.get('MoneroDaemon', {}).items()}
+        cfg_cloudflare = {'cf_%s' % k: try_cast(v) for k, v in config._sections.get('cloudflareAPI', {}).items()}
```

Comment:
There is an issue here when using --from-config arg. The config file seems changed for more DNS provider. The target of `config._sections` should be now `'DNS'`. But after changing that, I still got this error: `TypeError: __init__() got an unexpected keyword argument 'cf_provider'`.

Reply:
I think you are on an old version. https://github.com/skftn/moneriote-python/blob/refactor/moneriote/utils.py#L81 parses the ini.

Reply:
```
(venv) C:\Users\walass\Desktop\moneriote-python-refactor>moneriote.py --from-con
fig config.ini
• ▌ ▄ ·.        ▐ ▄ ▄▄▄ .▄▄▄  ▪        ▄▄▄▄▄▄▄▄ .
·██ ▐███▪▪     •█▌▐█▀▄.▀·▀▄ █·██ ▪     •██  ▀▄.▀·
▐█ ▌▐▌▐█· ▄█▀▄ ▐█▐▐▌▐▀▀▪▄▐▀▀▄ ▐█· ▄█▀▄  ▐█.▪▐▀▀▪▄
██ ██▌▐█▌▐█▌.▐▌██▐█▌▐█▄▄▌▐█•█▌▐█▌▐█▌.▐▌ ▐█▌·▐█▄▄
▌
▀▀  █▪▀▀▀ ▀█▄▀▪▀▀ █▪ ▀▀▀ .▀  ▀▀▀▀ ▀█▄▀▪ ▀▀▀  ▀▀▀

 @skftn / dsc
 @Lafudoci
 @gingeropolous
 @connorw600

Traceback (most recent call last):
  File "C:\Users\walass\Desktop\moneriote-python-refactor\moneriote.py", line 55
2, in <module>
    mon = Moneriote.from_config()
  File "C:\Users\walass\Desktop\moneriote-python-refactor\moneriote.py", line 28
4, in from_config
    return cls(**{**cfg_cloudflare, **cfg_monero})
TypeError: __init__() missing 4 required positional arguments: 'cf_domain_name',
 'cf_subdomain_name', 'cf_dns_api_key', and 'cf_dns_api_email'
```
It's from the last commit. I believe the issue is on my side (windows x64). But I can't figure it out. Could you help with this?

Reply:
You are defeinitely on an older commit because the banner is still in the old style. New one: [link](https://github.com/skftn/moneriote-python/blob/refactor/moneriote/utils.py#L10). Try to pull from skftn/moneriote-python:refactor

Reply:
Tried again and the commit is the same, but I found the behavior is different between cmd command `python moneriote.py` and `moneriote`. The former runs root dir's moneriote.py (with old banner) instead of utils.py (new banner). I guess I should use the latter as README metioned. But then I got error about `ModuleNotFoundError: No modual named 'rsa'.`
After I pip install rsa, I got:
`ModuleNotFoundError: No modual named 'suds'.`
When I try pip install suds, I got:
`ModuleNotFoundError: No modual named 'client'.`
After pip install client and try to run `moneriote --from-con
fig config.ini` Then I got:
```
(venv) C:\Users\LeftC\Downloads\moneriote-python>moneriote --from-config config.ini
Traceback (most recent call last):
  File "C:\Users\LeftC\Downloads\moneriote-python\venv\Scripts\moneriote-script.py", line 11, in <module>    load_entry_point('moneriote', 'console_scripts', 'moneriote')()
  File "C:\Users\LeftC\Downloads\moneriote-python\venv\lib\site-packages\click-7.0-py3.6.egg\click\core.py", line 764, in __call__    return self.main(*args, **kwargs)
  File "C:\Users\LeftC\Downloads\moneriote-python\venv\lib\site-packages\click-7.0-py3.6.egg\click\core.py", line 717, in main    rv = self.invoke(ctx)
  File "C:\Users\LeftC\Downloads\moneriote-python\venv\lib\site-packages\click-7.0-py3.6.egg\click\core.py", line 956, in invoke    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\LeftC\Downloads\moneriote-python\venv\lib\site-packages\click-7.0-py3.6.egg\click\core.py", line 555, in invoke    return callback(*args, **kwargs)
  File "c:\users\leftc\downloads\moneriote-python\moneriote\main.py", line 31, in cli    from moneriote.moneriote import Moneriote
  File "c:\users\leftc\downloads\moneriote-python\moneriote\moneriote.py", line 12, in <module>    from moneriote.dns import DnsProvider
  File "c:\users\leftc\downloads\moneriote-python\moneriote\dns\__init__.py", line 28, in <module>    from moneriote.dns.transip import TransIP
  File "c:\users\leftc\downloads\moneriote-python\moneriote\dns\transip.py", line 8, in <module>    from suds.client import Client as SudsClient
  File "C:\Users\LeftC\Downloads\moneriote-python\venv\lib\site-packages\suds\client.py", line 242
    except Exception, e:
                    ^
SyntaxError: invalid syntax
```
Please help :(

Reply:
First of all, thanks for trying :)

I think this will fix it:

1. Git pull (I made a commit)
2. Have your `config.ini` look like this:

```
[MoneroDaemon]
path = /home/dsc/Downloads/monero-gui-v0.12.3.0/monerod
address = 127.0.0.1
port = 18081
auth =
height_discovery_method = compare

[DNS]
provider = cloudflare
domain_name = example.com
subdomain_name = node
api_key = xxx
api_email = xxx
max_records = 5
```

With your details of course.

3. `moneriote --from-config config.ini`

It should now work :)